dhis2/dhis2-core
# feat: followup analysis with UIDs [DHIS2-10658]

### Summary
Makes the inputs and outputs of follow up analysis (which is a read-only request) accessible to a world of UIDs (instead of exposing and requiring internal numeric database table IDs).

Has all the features of the replaced ID version but also accepts more combinations of inputs (see javadoc) and does sharing access checks for the `DataElement`s. Also everything happens in the database instead of doing parts in the object model.

The request is validated to make sure all required parameters are present and reasonable.

### Automatic Tests
Added a component integration test that set ups a scenario using REST API and also requests the follow-up analysis.

Analysis has coverage for different filters in simple scenarios as well as validation.

Please also have a look at the documentation PR: https://github.com/dhis2/dhis2-docs/pull/688

### Manual Testing 
By the time QA test this the documentation changes should be effective here https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-master/data-validation.html

Things to test:
1. Validation:  not providing required parameters should return `CONFLICT` and a helpful error messge
2. Filtering: create a setup with some data values where `followup` is set to `true` for different element, org units, data sets or category option combos. Looking at https://github.com/dhis2/dhis2-core/blob/2590d99300f83a6c6ceee9393e7cb675db09b2bf/dhis-2/dhis-web/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/FollowupAnalysisControllerTest.java can give you an idea on how to create such test scenario. In essence this is simple filtering where the data values we want to see are those connected to the filter parameters we did provide.
